### PR TITLE
feat(datepicker): step 3, excluding dates

### PR DESCRIPTION
**Pull request by Leotheluck**
Adding options to disable specific dates of the Datepicker.

New enum: 
`ODS_DATEPICKER_DAY`: monday = 1, saturday = 6, etc.

New attributes:
> datesDisabled?: Date[] (specific dates to disable)
> daysOfWeekDisabled?: ODS_DATEPICKER_DAY[] (ex: monday, sunday disabled for all dates)
> maxDate?: Date (maximum date selectable)
> minDate?: Date (minimum date selectable)

### CODE REVIEW

File: packages/components/datepicker/src/components/osds-datepicker/core/controller.ts [hunk 1]
```diff
@@ -23,10 +23,17 @@ class OdsDatepickerController {
       if (newValue === undefined || newValue === null || isNaN(newValue.getTime())) {
```
Comment (dpellier) [inline packages/components/datepicker/src/components/osds-datepicker/core/controller.ts:23]: you can add also `|| !this.validateValue(newValue)` condition and remove the last else

File: packages/components/datepicker/src/components/osds-datepicker/core/controller.ts [hunk 2]
```diff
@@ -35,6 +42,26 @@ class OdsDatepickerController {
...
     this.component.emitBlur();
   }
+
+  private getMidnightDate(date: Date) {
+    return new Date(Date.UTC(date.getFullYear(), date.getMonth(), date.getDate()))
+  }
+
+  private validateValue(value?: Date) {
+    if (value) {
+      if (!this.component.maxDate || this.getMidnightDate(value).getTime() <= this.getMidnightDate(this.component.maxDate).getTime() ) {
```
Comment (dpellier) [inline packages/components/datepicker/src/components/osds-datepicker/core/controller.ts:52]: you should split your validation and avoid calling the same function (`getMidnightDate`) so many times, more like

```
if (!value) {
  return false;
}

const midnightDate = new Date(Date.UTC(date.getFullYear(), date.getMonth(), date.getDate()))

if (this.component.maxDate) {
  return <test max date>
}

if (this.component.minDate) {
  return <test max date>
}

...
```

**Review by dpellier — CHANGES_REQUESTED** [review]

File: packages/components/datepicker/src/components/osds-datepicker/core/controller.ts (hunk 1 above)
Comment (aesteves60) [reply packages/components/datepicker/src/components/osds-datepicker/core/controller.ts:23]: Can you create a function for add context like `isValidValue`

File: packages/storybook/stories/components/datepicker/datepicker.web-components.stories.ts [hunk 3]
```diff
@@ -38,6 +44,16 @@ const storyParams = {
     category: 'General',
     defaultValue: false,
   },
+  maxDate: {
+    category: 'General',
+    defaultValue: null,
+    options: [null, new Date('1999-11-02'), new Date('2024-11-02')],
+  },
+  minDate: {
```
Comment (aesteves60) [inline packages/storybook/stories/components/datepicker/datepicker.web-components.stories.ts:52]: I have an error on storybook with the date
image
And when I select one the input not display the value

**Review by aesteves60 — CHANGES_REQUESTED** [review]

File: packages/components/datepicker/src/components/osds-datepicker/osds-datepicker.tsx [hunk 4]
```diff
@@ -66,9 +79,32 @@ export class OsdsDatepicker implements OdsDatepickerAttribute, OdsDatepickerEven
...
+    }
+  }
+
+  @Watch('datesDisabled')
+  @Watch('daysOfWeekDisabled')
+  @Watch('format')
+  @Watch('maxDate')
+  @Watch('minDate')
+  updateDatepicker() {
+    if (this.datepickerInstance) {
```
Comment (aesteves60) [inline packages/components/datepicker/src/components/osds-datepicker/osds-datepicker.tsx:94]: Early return for less tabulation

**Review by aesteves60 — CHANGES_REQUESTED** [review]

**Review by astagnol — CHANGES_REQUESTED** [review]
Capture d’écran 2023-09-28 à 11 56 34

In Storybook, it can be a bit confusing for users when they choose a min and/or max date and the selected value is not displayed. Still, it's working as expected but I think we should display the selected value in select.

**Review by astagnol — APPROVED** [review]

**Review by aesteves60 — APPROVED** [review]

**Pull request merged**